### DISCUSSION TITLE
Limited styling options

### DISCUSSION THREAD

**Discussion by raphaelbernhart** [dargmuesli/nuxt-cookie-control#228] [Ideas]
Hi,

I am thinking about using this module in production but for this to work for me IMO the options for styling are very limited.

Please correct me if I'm wrong but aren't the only options to change the styling:
1. The colors in the nuxt config
2. classes in the component directly (f.e. via deep modifier)

But then the questions comes to me, why don't all the children in the component have a class. ATM it is really difficult to style certain elements in the component tree. [image](https://github.com/user-attachments/assets/c3245c8f-4b02-43ba-a138-c197c7df31fe).
Would probably also be nice to be able to add custom classes. This would be useful f.e. when using tailwindcss.

FYI: I am open to create a PR, I just want to make sure that I am on the same page as everyone else :)

+ Maybe a feature to collapse the Cookie groups? When there are a lot of cookies IMO it gets really hard for the user to keep track. This probably only makes sense once the feature in [this](https://github.com/dargmuesli/nuxt-cookie-control/issues/216) issue is implemented? As now, for a workaround you can just "group" the cookies inside of the functional and optional cookies groups. [image](https://github.com/user-attachments/assets/c8e20d9c-58fb-4afc-8574-f44fd85dc8a6)

**Comment by dargmuesli** [comment]
Yes, I definitely long for a tailwindcss implementation as well, just haven't found the time for that. Also, subcomponents seem to make sense to me too! It might help with feature requests like this one too: https://github.com/dargmuesli/nuxt-cookie-control/issues/196

**Reply by raphaelbernhart** [reply]
Sorry what do you mean with subcomponents? The collapse feature?

**Reply by dargmuesli** [reply]
Right now there is mainly only the `CookieControl` component, e.g.
<img src="https://github.com/user-attachments/assets/a8592eb8-f5c3-4303-9cf2-8945c1db7a7f" width="250">

That component's template could be more modularized, e.g. to have a `CookieControlElement` subcomponent for more flexible layout options:
<img src="https://github.com/user-attachments/assets/4903c88a-92ee-4334-90a2-53f1954a64a7" width="250">
:

**Reply by raphaelbernhart** [reply]
> Right now there is mainly only the `CookieControl` component, e.g. <img alt="" width="250" src="https://private-user-images.githubusercontent.com/4778485/365226572-a8592eb8-f5c3-4303-9cf2-8945c1db7a7f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU2NDE4MDUsIm5iZiI6MTcyNTY0MTUwNSwicGF0aCI6Ii80Nzc4NDg1LzM2NTIyNjU3Mi1hODU5MmViOC1mNWMzLTQzMDMtOWNmMi04OTQ1YzFkYjdhN2YucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDZUMTY1MTQ1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmEyM2FlM2VmOGE0MzE2MGQ4ODZhY2ZiMjYzZmUxNWM4NTM2NDAxOThhYTE3MDNlNWQ2OWU1NmE0YjJiMGYxNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.cN62KwZZj9yyVHLEGgfhWklpSv7olEvJAP5o-ga0bFE">
> 
> That component's template could be more modularized, e.g. to have a `CookieControlElement` subcomponent for more flexible layout options: <img alt="" width="250" src="https://private-user-images.githubusercontent.com/4778485/365226667-4903c88a-92ee-4334-90a2-53f1954a64a7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU2NDE4MDUsIm5iZiI6MTcyNTY0MTUwNSwicGF0aCI6Ii80Nzc4NDg1LzM2NTIyNjY2Ny00OTAzYzg4YS05MmVlLTQzMzQtOTBhMi01M2YxOTU0YTY0YTcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDZUMTY1MTQ1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzFiMGRhYmU0ZDQyMzg0ZDVkOGRjZjE2ZmY3YTVhY2NmNGQzOTc2MWY0ZjIyYzU3ZmEyMzViM2RjNDY4NzM0MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.mNzhVckruxx76a66EyhOT9WDXpEVZmyZyJ7FHJGDgzo"> :

Ah ok yeah. Tho IMO those are two separat features. I'll have a look in the upcoming days :)

**Comment by dargmuesli** [comment]
> why don't all the children in the component have a class

I just want to mention that you should of course be able to access DOM children of components with classes via the usual CSS syntax like `.cookieControl__ModalContent input:checked+button`

**Comment by raphaelbernhart** [comment]
@dargmuesli As there are probably more things, can we continue the discussion here instead of the issue (https://github.com/dargmuesli/nuxt-cookie-control/issues/229)?

So there are probably these things:
1. Re-structure the component (using sub-components)
2. Config to be capable to add custom styling (preferably also tailwindcss)
3. add disable-scroll feature (maybe with use of headlessui)

Reg. "1." - I think we could change the structure (which we prob. will have to if we choose to use a component library) or we try to get "closest to original". This would definitely help with compatibility issues.

Reg. "2." - I think the easiest way would be to add a "styles" (?) config prop where you can update the styles of the elements of the cookie control? Do you agree/have other ideas?

Reg. "3." - If we decide to not use a third-party I think we only have to keep in mind that iOS has some caveats + we should not use `body-scroll-lock` as it seems to be abandoned (maybe use https://ishadeed.com/article/prevent-scroll-chaining-overscroll-behavior/?)

**Reply by dargmuesli** [reply]
> I think the easiest way would be to add a "styles" (?) config prop where you can update the styles of the elements of the cookie control? Do you agree/have other ideas?

I'd just want the styles to not end up inline (except if Nuxt is configured to do that by downstream users), everything else is fine for me. Maybe also check how https://ui.nuxt.com/ offers configuration. I think they make use of the `appconfig`, which we could or should do too maybe?

> add disable-scroll feature (maybe with use of headlessui)

The point of headlessui is more the headless part instead of the ui as I understand it. So it helps with all the accessibility stuff and maybe also the usual stuff like scroll locking that doesn't need to be reinvented for each project or library, which could very much help here.

**Reply by raphaelbernhart** [reply]
> The point of headlessui is more the headless part instead of the ui as I understand it.

True but tho when exchanging, even if only the modal, we would inevitably have a different structure.
Anyways, I'll give it a shot and see how it turns out.

**Reply by dargmuesli** [reply]
Hey @raphaelbernhart, did you achieve some success? :)

**Reply by raphaelbernhart** [reply]
> Hey @raphaelbernhart, did you achieve some success? :)

Hi, I did unfortunately not have time to work further on it :/
